src/execution/operator/csv_scanner/sniffer/dialect_detection.cpp

```diff
@@ -0,0 +1,311 @@
+#include "duckdb/execution/operator/persistent/csv_scanner/csv_sniffer.hpp"
+#include "duckdb/main/client_data.hpp"
+
+namespace duckdb {
+
+struct SniffDialect {
+	inline static void Initialize(CSVStateMachine &machine) {
+		machine.state = CSVState::STANDARD;
+		machine.previous_state = CSVState::STANDARD;
+		machine.pre_previous_state = CSVState::STANDARD;
+		machine.cur_rows = 0;
+		machine.column_count = 1;
+	}
+
+	inline static bool Process(CSVStateMachine &machine, vector<idx_t> &sniffed_column_counts, char current_char) {
+
+		D_ASSERT(sniffed_column_counts.size() == machine.options.sample_chunk_size);
+
+		if (machine.state == CSVState::INVALID) {
+			sniffed_column_counts.clear();
+			return true;
+		}
+		machine.pre_previous_state = machine.previous_state;
+		machine.previous_state = machine.state;
+
+		machine.state = static_cast<CSVState>(
+		    machine.transition_array[static_cast<uint8_t>(machine.state)][static_cast<uint8_t>(current_char)]);
+		bool empty_line =
+		    (machine.state == CSVState::CARRIAGE_RETURN && machine.previous_state == CSVState::CARRIAGE_RETURN) ||
+		    (machine.state == CSVState::RECORD_SEPARATOR && machine.previous_state == CSVState::RECORD_SEPARATOR) ||
+		    (machine.state == CSVState::CARRIAGE_RETURN && machine.previous_state == CSVState::RECORD_SEPARATOR) ||
+		    (machine.pre_previous_state == CSVState::RECORD_SEPARATOR &&
+		     machine.previous_state == CSVState::CARRIAGE_RETURN);
```

Comment:
If within reasonable effort, we should have this as another state in the state machine. We can then skip `// It means our carriage return is a record separator` and `// Identify what is our line separator`, as that should happen when we try to find the best state machine (I think 😄). `Finalize` should also become simpler.